### DISCUSSION TITLE
Use environment variables to template repository config "{{ env.MYVAR }}"

### DISCUSSION THREAD

**Discussion by luckyf** [renovatebot/renovate#17175] [Request Help]
Hi,

in a self-hosted renovate instance I am currently using the [ `secrets`](https://docs.renovatebot.com/self-hosted-configuration/#secrets) setting to template values of the repository settings.
This works fine but as it's correct for secrets they get redacted from Pull Requests and logs.
As the `NPM_SCOPE` variable doesn't really represent a secret and therefore shouldn't be redacted, is there a way to use for example environment variables for the templating inside the repository config?

Some syntax like `{{ env.NPM_SCOPE }}` would be really useful in combination with the [`customEnvVariables`](https://docs.renovatebot.com/self-hosted-configuration/#customenvvariables) setting.

Self-hosted config:
```js
module.exports = {
  // ...
  secrets: {
    API_TOKEN: 'xxxx',
    NPM_SCOPE: 'projectname'
  },
  // ...
};
```

Repository config:
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "packageRules": [
    {
      "excludePackagePatterns": [
        "^@{{ secrets.NPM_SCOPE }}"
      ]
    }
  ],
}
```

**Comment by viceice** [comment]
no

**Reply by luckyf** [reply]
Hi @viceice,
thanks for your fast answer.
May I create a Feature Request for this?

Using secrets would be ok if there would be only cosmetic issues but unfortunately it is also causing `CONFIG_SECRETS_EXPOSED` errors during running.
The Renovate Bot throws errors with the message `Repository aborted due to potential secrets exposure` and then cancels the repository.

**Reply by viceice** [reply]
It's not as easy, as we can't expose all env because of security reasons

**Reply by luckyf** [reply]
Good point.
Maybe using the same mechanism as for secrets and not env variables would be a suitable idea?
The functionality could be almost exactly the same as for secrets, just without the redated scan.
It would enable the Renovate-Admin to provide variables without exposing the environment variables.

Maybe something like:
self-hosted config
```js
module.exports = {
  // ...
  secrets: {
    API_TOKEN: 'xxxx',
  },
  vars: {
    NPM_SCOPE: 'projectname'
  },
  // ...
};
```

Repository config:
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "packageRules": [
    {
      "excludePackagePatterns": [
        "^@{{ vars.NPM_SCOPE }}"
      ]
    }
  ],
}
```

**Reply by rarkins** [reply]
Why do you need to read `NPM_SCOPE` from the environment instead of putting it into Renovate config? If it's not a secret, why put it into secrets?

**Reply by luckyf** [reply]
@rarkins I am trying to make the repo config as generic as possible to use it for multiple projects and then inject the project specific parts (like NPM scope or private registry URL) via CI-Pipeline-Job into the renovate self hosted config.

**Reply by rarkins** [reply]
Do you have one pipeline running per repo? e.g. instead of a single pipeline servicing dozens or hundreds of repos, each repo uses their own pipeline for Renovate?

**Reply by luckyf** [reply]
No. I only run one scheduled renovate pipeline and activated autodiscovery for the services/packages group in the "tools/renovate-bot" repository.
As a workaround I hardcoded the project specific variables into the shareable config preset but for me it would be more scalable to use one preset for all projects and only update the pipeline config of the renovate-bot.


I use Gitlab Groups to structure the repositories for each project like this:
```
config/
├─ renovate         <= Shareable config preset (JSON repository config)
services/
├─ service1         <= Extends sharable config preset  with custom config (e.g. other assignee per service) JSON
├─ service2         <= Extends sharable config preset  with custom config (e.g. other assignee per service) JSON
├─ ...
packages/
├─ package1         <= Extends sharable config preset with custom config (e.g. other assignee per service) JSON
├─ package2         <= Extends sharable config preset with custom config (e.g. other assignee per service) JSON
├─ ...
tools/
├─ renovate-bot     <= Includes self-hosted renovate config and Pipeline runs only on this repository (JS config file)
...
```

**Reply by luckyf** [reply]
With and additional `vars` object, the configuration could look like this.

Self-hosted config (config.js):
This file could be generic and I could use it for all projects. The environment variables are injected during the pipeline run and set for each project separatly.
```js
module.exports = {
  // ...
  secrets: {
    API_TOKEN: process.env.API_TOKEN,
  },
  vars: {
    NPM_SCOPE: process.env.NPM_SCOPE,
  },
  // ...
};
```

Sharable preset (renovate-preset.json)
This file could be generic too.
```json
{
  ...
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "hostRules": [
    {
      "hostType": "npm",
      "matchHost": "{{ vars.CI_API_V4_URL }}/projects/{{ vars.NPM_REGISTRY_PROJECTID }}/packages/npm/",
      "token": "{{ secrets.BOT_PASSWORD }}"
    }
  ],
  "packageRules": [
    {
      "excludePackagePatterns": [
        "^@{{ vars.NPM_SCOPE }}"
      ]
    }
  ],
  ...
}
```


Repo config for each service/package (renovate.json):
This file extends the global preset for this project and sets service specific values (e.g. the reviewer who is responsible for this service)
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "enabled": true,
  "extends": [
    "local>project/config/renovate:renovate-preset"
  ],
  "reviewers": [
    "otheruser"
  ],
  ...
}
```

It also would enable to separate concerns between the "renovate-Admin" who maintains the `tools/renovate-bot` repository and sets `secrets` and `vars` and the "developer" who manages the default update behaviour by maintaining the `config/renovate` repository like we already can do for secrets.

The only difference between the `secrets` and `vars` object would be that the `vars` don't get redacted and don't trigger `CONFIG_SECRETS_EXPOSED` errors.
